### PR TITLE
[Concurrency] Introduce Actor protocol to which actor classes all conform

### PR DESCRIPTION
**Pull request by DougGregor**
Introduce a new `Actor` protocol, which is a class-bound protocol with only
one requirement:

```
    func enqueue(partialTask: PartialAsyncTask)
```

All actor classes implicitly conform to this protocol, and will synthesize
a definition of `enqueue(partialTask:)` along with storage for the runtime
to manager in the actor.

This design also lets us loosen the constraints on actors somewhat, so that
an actor class can be `@objc` and inherit directly from `NSObject`. Allow this,
and tighten up `@objc` checking to maintain actor isolation across the
language boundary.

### CODE REVIEW

File: lib/Sema/DerivedConformanceActor.cpp [hunk 1]
```diff
@@ -0,0 +1,104 @@
+//===--- DerivedConformanceEquatableHashable.cpp - Derived Equatable & co -===//
```
Comment (theblixguy) [inline lib/Sema/DerivedConformanceActor.cpp:1]: Nit:
```suggestion
//===--- DerivedConformanceActor.cpp - Derived Actor conformance -----------===//
```

Comment (DougGregor) [reply lib/Sema/DerivedConformanceActor.cpp:1]: Ah, thanks. I made the change locally.

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Pull request merged**